src/modules/imageresizer/dll/ContextMenuHandler.cpp

```diff
@@ -0,0 +1,407 @@
+// ContextMenuHandler.cpp : Implementation of CContextMenuHandler
+
+#include "stdafx.h"
+#include "ContextMenuHandler.h"
+#include "HDropIterator.h"
+#include "Settings.h"
+#include "common/icon_helpers.h"
+#include "trace.h"
+
+extern HINSTANCE g_hInst_imageResizer;
+
+CContextMenuHandler::CContextMenuHandler()
+{
+    m_pidlFolder = NULL;
+    m_pdtobj = NULL;
+    app_name = GET_RESOURCE_STRING(IDS_RESIZE_PICTURES);
+}
+
+CContextMenuHandler::~CContextMenuHandler()
+{
+    Uninitialize();
+}
+
+void CContextMenuHandler::Uninitialize()
+{
+    CoTaskMemFree((LPVOID)m_pidlFolder);
+    m_pidlFolder = NULL;
+
+    if (m_pdtobj)
+    {
+        m_pdtobj->Release();
+        m_pdtobj = NULL;
+    }
+}
+
+HRESULT CContextMenuHandler::Initialize(_In_opt_ PCIDLIST_ABSOLUTE pidlFolder, _In_opt_ IDataObject* pdtobj, _In_opt_ HKEY hkeyProgID)
+{
+    Uninitialize();
+
+    if (!CSettings::GetEnabled())
+    {
+        return E_FAIL;
+    }
+
+    if (pidlFolder)
+    {
+        m_pidlFolder = ILClone(pidlFolder);
+    }
+
+    if (pdtobj)
+    {
+        m_pdtobj = pdtobj;
+        m_pdtobj->AddRef();
+    }
+
+    return S_OK;
+}
+
+HRESULT CContextMenuHandler::QueryContextMenu(_In_ HMENU hmenu, UINT indexMenu, UINT idCmdFirst, UINT idCmdLast, UINT uFlags)
+{
+    if (uFlags & CMF_DEFAULTONLY)
+    {
+        return S_OK;
+    }
+    if (!CSettings::GetEnabled())
+    {
+        return E_FAIL;
+    }
+    // NB: We just check the first item. We could iterate through more if the first one doesn't meet the criteria
+    HDropIterator i(m_pdtobj);
+    i.First();
+// Suppressing C26812 warning as the issue is in the shtypes.h library
+#pragma warning(suppress : 26812)
+    PERCEIVED type;
+    PERCEIVEDFLAG flag;
+    LPTSTR pszPath = i.CurrentItem();
+    LPTSTR pszExt = PathFindExtension(pszPath);
+
+    // TODO: Instead, detect whether there's a WIC codec installed that can handle this file
```

Comment:
I haven't seen any noticeable performance implication of this so far. To avoid a performance issue with increasing number of images, I've added the check only for the file on which the right click is done (which is the first file in the list). ImageResizer has appropriate error messages if you trying resizing a "non-image" file and it still makes sure all the images in the selection list do get resized without crashing midway because of the wrong files.